content/backend/graphql-scala/2-preparing-first-query.md

```diff
@@ -0,0 +1,377 @@
+---
+title: Preparing to first query
+pageTitle: "The last steps of preparation to able run first GraphQL query."
+description: "In this chapter you will learn last steps of preparation, include graphiql console for easy debugging and at the end run your first query."
+question: "Does GraphQL needs HTTP Server?"
+answers: ["Yes. It needs HTTP server.", "Yes, it needs HTTP server but some of features can be used without that", "No, but it strictly recommended to use. Without HTTP layer, GraphQL is losing some of its features.","No, GraphQL is specification is far away from tranportation protocol. You can use HTTP, Websockets, sockets or even use it internally in you application." ]
+correctAnswer: 3
+---
+
+### Goal
+
+Our goal for this chapter is to run following query:
+
+```graphql
+
+query {
+  allLinks {
+    id
+    name
+    description
+  }
+}
+```
+The expected result is a list of few links.
+
+### Define a model
+
+Let's start from defining a really basic model
+
+<Instruction>
+
+Create class `Link`:
+
+```scala
+
+case class Link(id: Int, url: String, description: String)
+
+```
+
+</Instruction>
+
+As you can see, a `Link` model has less fields than in the Schema You saw in the first chapter, but no worries, we will improve the model in the future. Now we're focusing on completing execution stack so it would be better to keep this model simple.
+
+<Instruction>
+
+Add following chanages to the `DBSchema.scala`:
+
+```scala
+
+//1
+class LinksTable(tag: Tag) extends Table[Link](tag, "LINKS"){
+
+    def id = column[Int]("ID", O.PrimaryKey)
+    def url = column[String]("URL")
+    def description = column[String]("DESCRIPTION")
+
+    def * = (id, url, description) <> ((Link.apply _).tupled, Link.unapply)
+
+}
+
+//2
+val Links = TableQuery[LinksTable]
+
+//3
+val databaseSetup = DBIO.seq(
+    Links.schema.create,
+
+    Links ++= Seq(
+      Link(0, "http://howtographql.com", "Awesome community driven GraphQL tutorial"),
+      Link(0, "http://graphql.org", "Official GraphQL webpage"),
+      Link(0, "https://facebook.github.io/graphql/", "GraphQL specification")
+    )
+)
+```
+
+</Instruction>
+
+We just added database definition of our first model.
+* **1** defines mapping to database table,
+* **2** gives us an helper we will use to accessing data in this table.
+* The last change, **3** is responsible for creating schema and adding three entities to the database.
+
+### Context
+
+Context is an object that flows across the whole execution, in the most cases this object doesn't change at all. The main responsibility of the Context it's to provide all the things that helps with fulfilling. In our example we will put there `DAO` so all queries will have access to the database, in the future we will also put there authentication data.
```

Comment:
`it's to provide all the things that helps with fulfilling` -> `is to provide data and utils needed to fulfill the query`